samples/nrf9160/aws_fota/Kconfig

```diff
@@ -36,35 +37,53 @@ config PROVISION_CERTIFICATES
 	  certificates with anyone.
 
 config USE_CLOUD_CLIENT_ID
+	depends on !USE_NRF_CLOUD
 	bool "Custom MQTT client ID"
 	default y
-if USE_CLOUD_CLIENT_ID
+
 config CLOUD_CLIENT_ID
+	depends on !USE_NRF_CLOUD
+	depends on USE_CLOUD_CLIENT_ID
 	string "Client ID"
 	default "your_client_id"
-endif
 
 config MQTT_BROKER_HOSTNAME
+	depends on !USE_NRF_CLOUD
 	string "AWS IoT MQTT broker hostname"
 	default "your_aws_mqtt_broker_hostname.amazonaws.com"
 	help
 	  Default is set to be the nRF Cloud MQTT broker.
+
 config MQTT_BROKER_PORT
+	depends on !USE_NRF_CLOUD
 	int "AWS IoT MQTT broker port"
 	default 8883
-endif #!USE_NRF_CLOUD
 
-if USE_NRF_CLOUD
 config MQTT_BROKER_HOSTNAME
+	depends on USE_NRF_CLOUD
 	string
 	default "a2n7tk1kp18wix-ats.iot.us-east-1.amazonaws.com"
 	help
 	  Default is set to be the nRF Cloud MQTT broker.
 
 config MQTT_BROKER_PORT
+	depends on USE_NRF_CLOUD
 	int
 	default 8883
-endif #USE_NRF_CLOUD
+
+config NRF_CLOUD_CLIENT_ID_PREFIX
+	depends on !USE_CLOUD_CLIENT_ID
```

Comment:
Not exactly sure what your intentions are.
Here you have a dependency to `!USE_CLOUD_CLIENT_ID` but that doesn't remove the configuration option, cause it is still accessible here:
https://github.com/nrfconnect/sdk-nrf/blob/9d31e898c6125c672881729897a471c273d491bf/subsys/net/lib/nrf_cloud/Kconfig#L40-L47

which leads to next question.
Why is this setting depending on `!NRF_CLOUD` when the code referenced iin the link has a dependency that `NRF_CLOUD=y` ?

See also this screenshot:
![image](https://user-images.githubusercontent.com/9802583/94817559-6b7e9880-03fd-11eb-9beb-3ddf4d6748a4.png)

Reply:
This is due to that the FOTA library can be used without the nRF Connect for Cloud library, but still connect to nRF Connect for Cloud broker.

Note that it depends on  **!NRF_CLOUD** (the config value enabling the nrf_cloud module), not **!USE_NRF_CLOUD**. In that case provide the `NRF_CLOUD_CLIENT_ID_PREFIX` used in the FOTA library through its own config.

Reply:
> not !USE_NRF_CLOUD. 

Sorry for the confusion, I see I posted screenshot of wrong menu entry.
I never meant to refer to `USE_NRF_CLOUD`.
This should be correct:
![image](https://user-images.githubusercontent.com/9802583/94849370-97ae0f80-0425-11eb-870d-99cbaaba0eb5.png)


If you read the code here:
https://github.com/nrfconnect/sdk-nrf/blob/9d31e898c6125c672881729897a471c273d491bf/samples/nrf9160/aws_fota/Kconfig#L74-L76
then you  have `NRF_CLOUD_CLIENT_ID_PREFIX ` available when `!USE_CLOUD_CLIENT_ID && !NRF_CLOUD` which I describe and show in the screenshot.
That is when `USE_CLOUD_CLIENT_ID=n && NRF_CLOUD=n`.

But this code:
https://github.com/nrfconnect/sdk-nrf/blob/9d31e898c6125c672881729897a471c273d491bf/subsys/net/lib/nrf_cloud/Kconfig#L40
is guarded by this:
https://github.com/nrfconnect/sdk-nrf/blob/9d31e898c6125c672881729897a471c273d491bf/subsys/net/lib/nrf_cloud/Kconfig#L13

which means that `NRF_CLOUD_CLIENT_ID_PREFIX` is available when `NRF_CLOUD=y`.

So actually, in this sample, the only time `NRF_CLOUD_CLIENT_ID_PREFIX` is NOT available is when `USE_CLOUD_CLIENT_ID=y && NRF_CLOUD=n`.
Not sure if that is your intention, but it looks looks susspicious to me.

But if it is correct behavior, then you could write the code as:

>config NRF_CLOUD_CLIENT_ID_PREFIX
>	depends on !USE_CLOUD_CLIENT_ID
>	~depends on !NRF_CLOUD~ (can be removed)

as that would allow user to specify prefix in both places when `NRF_CLOUD=y`.
If your intention is to force the user to go into `Kernel -> Modules -> Nordic nRF Connect -> Networking -> Application protocols -> nRF Cloud Library` when `USE_CLOUD_CLIENT_ID=y` and `NRF_CLOUD=y` then code is correct. (Although I don't see why user cannot have the setting accessible in both places in the case)

I just want to ensure that there is not a flaw in the Kconfig dependencies here, resulting in user being able to specify `NRF_CLOUD_CLIENT_ID_PREFIX=<prefix>`, when he shouldn't have that option.

Reply:
> So actually, in this sample, the only time `NRF_CLOUD_CLIENT_ID_PREFIX` is NOT available is when 
> `USE_CLOUD_CLIENT_ID=y && NRF_CLOUD=n`.
> Not sure if that is your intention, but it looks looks susspicious to me.

Yes, this is intentional. If `USE_CLOUD_CLIENT_ID=y` the MQTT client Id is taken from `CLOUD_CLIENT_ID` and is not constructed using `${NRF_CLOUD_CLIENT_ID_PREFIX}-${IMEI}`. 

> If your intention is to force the user to go into `Kernel -> Modules -> Nordic nRF Connect -> Networking -> Application protocols -> nRF Cloud Library` when `USE_CLOUD_CLIENT_ID=y` and `NRF_CLOUD=y` then code is correct. (Although I don't see why user cannot have the setting accessible in both places in the case)

Yes, I wanted to have the setting only in one place, because I *assumed* that it could overwrite each other. I Kconfig allows to specify the same setting in multiple places then there is no harm in having it in both places.

Is this safe to do? If yes, I am fine with removing the `depends on !NRF_CLOUD` for `config NRF_CLOUD_CLIENT_ID_PREFIX`,

Reply:
> > Not sure if that is your intention, but it looks looks susspicious to me.
> 
> Yes, this is intentional. If `USE_CLOUD_CLIENT_ID=y` the MQTT client Id is taken from `CLOUD_CLIENT_ID` and is not constructed using `${NRF_CLOUD_CLIENT_ID_PREFIX}-${IMEI}`.

So in that case, if users does:
```
CONFIG_USE_CLOUD_CLIENT_ID=y
CONFIG_NRF_CLOUD=y
CONFIG_NRF_CLOUD_CLIENT_ID_PREFIX="myprefix"
```
then `myprefix` is disregarded, even though the setting is available for configuration under:
`Kernel -> Modules -> Nordic nRF Connect -> Networking -> Application protocols -> nRF Cloud Library` ?

Or is it still being used, but for something else.

> Yes, I wanted to have the setting only in one place, because I assumed that it could overwrite each other.

You can overwrite the default setting.
As example, for the `aws_fota` you could decide to use `aws-` as default prefix, while for any other sample that doesn't use `aws_fota/Kconfig` the default would be `nrf-`
 
>I Kconfig allows to specify the same setting in multiple places then there is no harm in having it in both places.

Yes, that is possible, and if user changes `nrf-` under `Kernel -> Modules -> Nordic nRF Connect -> Networking -> Application protocols -> nRF Cloud Library` then the new value will be visible under `MQTT FOTE Sample Settings`, and vice versa.

Reply:
As example:
![image](https://user-images.githubusercontent.com/9802583/94908706-66275980-04a2-11eb-8582-3fa53b35abe4.png)

Reply:
> So in that case, if users does:
> 
> ```
> CONFIG_USE_CLOUD_CLIENT_ID=y
> CONFIG_NRF_CLOUD=y
> CONFIG_NRF_CLOUD_CLIENT_ID_PREFIX="myprefix"
> ```
> 
> then `myprefix` is disregarded, even though the setting is available for configuration under:
> `Kernel -> Modules -> Nordic nRF Connect -> Networking -> Application protocols -> nRF Cloud Library` ?

In this case `myprefix` is disregarded in the `aws_fota` sample (it will use the value of `CLOUD_CLIENT_ID`) when configuring the cloud_client: https://github.com/nrfconnect/sdk-nrf/blob/4ba4b1e839454c00e336dff6bed4fcf87d2e0a28/samples/nrf9160/aws_fota/src/main.c#L382

> Or is it still being used, but for something else.

No, because here `NRF_CLOUD=n` so the `nrf_cloud` module will not be used.

And now we are coming back to why we need `CONFIG_NRF_CLOUD_CLIENT_ID_PREFIX` in two places: in case `NRF_CLOUD=n`, the setting will not be available: https://github.com/coderbyheart/sdk-nrf/blob/9d31e898c6125c672881729897a471c273d491bf/subsys/net/lib/nrf_cloud/Kconfig#L13-L47

Reply:
> > Or is it still being used, but for something else.
> 
> No, because here `NRF_CLOUD=n` so the `nrf_cloud` module will not be used.

But user can still set `NRF_CLOUD=y` and have `USE_CLOUD_CLIENT_ID=y`.
Would that be a mistake ?
Should `USE_CLOUD_CLIENT_ID` perhaps depend on `!NRF_CLOUD` ?
Like
```
config USE_CLOUD_CLIENT_ID
	depends on !USE_NRF_CLOUD
	depends on !NRF_CLOUD
```


>  in case `NRF_CLOUD=n`, the setting will not be available

Correct, but then it would be enough to just depend on `!USE_CLOUD_CLIENT_ID`, like I wrote in this message https://github.com/nrfconnect/sdk-nrf/pull/2961#discussion_r498446347:
That is:
```
config NRF_CLOUD_CLIENT_ID_PREFIX
	depends on !USE_CLOUD_CLIENT_ID
	string "Prefix used when constructing the MQTT client ID from the IMEI"
	default "nrf-"
	help
	  ......
```

Thanks for taking your time.
Just trying to discover if there are some errors in the dependency handling.

Reply:
> But user can still set `NRF_CLOUD=y` and have `USE_CLOUD_CLIENT_ID=y`.
> Would that be a mistake ?

No. This is if the user wants to use custom client ID ([source](https://github.com/nrfconnect/sdk-nrf/blob/ac09cf9a97f9b1f5f93741fa1cdba9979e47ae67/samples/nrf9160/aws_fota/README.rst)). A typical scenario is that they are running the `aws_fota` sample against their own AWS IoT instance.

```
.. option:: USE_CLOUD_CLIENT_ID - Custom MQTT client ID

   The client ID links your board to the thing in AWS IoT.
   By default, the client ID is ``nrf-IMEI``, where *IMEI* is the IMEI number of your board.
   If you chose a different name for your thing in AWS IoT, check this option and specify the AWS IoT thing name as client ID.
```

> Should `USE_CLOUD_CLIENT_ID` perhaps depend on `!NRF_CLOUD` ?

Well, it's unlikely, but possible to use `nrf_cloud` library to connect to the nRF Connect for Cloud broker and use the `aws_fota` library to connect **to a different** MQTT broker. So I would not force this dependency.

Reply:
Great, thanks for the explanation.

Whether or not `NRF_CLOUD_CLIENT_ID_PREFIX` should be shown under `MQTT FOTE Sample` when `CLOUD_CLIENT_ID=n` and `NRF_CLOUD=y` is a matter of preference that I leave for you to decide.
Afterall, the config option is available under `Kernel -> Modules -> Nordic nRF Connect -> Networking -> Application protocols -> nRF Cloud Library` when `NRF_CLOUD=y`